From a16f0356a9699fa49846cceb000b5f1d98a36c4e Mon Sep 17 00:00:00 2001
From: wu xia <xia.wu@intel.com>
Date: Tue, 12 Mar 2013 16:01:52 +0800
Subject: sqlite: enable ATOMIC_WRITE feature for sqlite lib

ATOMIC_WRITE feature will improve SQLite performance, especially for
small updates(data size is less than one page (4KB)).

Issue: BZ: 90764
Change-Id: I2019f9c77da2867eaf60e915b94b516b58f9e666
Signed-off-by: wu xia <xia.wu@intel.com>

Category: AOSP Improvement
Domain: AOSP.Optimization
Origin: Internal
Upstream-Candidate: No, proprietary
---
 dist/Android.mk |    1 +
 dist/sqlite3.c  |   14 +++++++++++---
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/dist/Android.mk b/dist/Android.mk
index 3e39229..fa76eae 100644
--- a/dist/Android.mk
+++ b/dist/Android.mk
@@ -24,6 +24,7 @@ common_sqlite_flags := \
 	-DSQLITE_ENABLE_FTS3 \
 	-DSQLITE_ENABLE_FTS3_BACKWARDS \
 	-DSQLITE_ENABLE_FTS4 \
+	-DSQLITE_ENABLE_ATOMIC_WRITE \
 	-DSQLITE_OMIT_BUILTIN_TEST \
 	-DSQLITE_OMIT_COMPILEOPTION_DIAGS \
 	-DSQLITE_OMIT_LOAD_EXTENSION \
diff --git a/dist/sqlite3.c b/dist/sqlite3.c
index c6c8a5e..d454e32 100644
--- a/dist/sqlite3.c
+++ b/dist/sqlite3.c
@@ -28673,11 +28673,19 @@ static int unixSectorSize(sqlite3_file *pFile){
 */
 static int unixDeviceCharacteristics(sqlite3_file *id){
   unixFile *p = (unixFile*)id;
+  int returnValue = 0;
+
   if( p->ctrlFlags & UNIXFILE_PSOW ){
-    return SQLITE_IOCAP_POWERSAFE_OVERWRITE;
-  }else{
-    return 0;
+    returnValue = SQLITE_IOCAP_POWERSAFE_OVERWRITE;
   }
+  // Begin Android change
+  // Enable ATOMIC_WRITE feature by default
+  #ifdef SQLITE_ENABLE_ATOMIC_WRITE
+  returnValue |= SQLITE_IOCAP_ATOMIC4K;
+  #endif
+  // End Android change
+
+  return returnValue;
 }
 
 #ifndef SQLITE_OMIT_WAL
-- 
1.7.4.1

